crates/rust-analyzer/src/handlers.rs

```diff
@@ -804,6 +804,41 @@ pub(crate) fn handle_runnables(
     Ok(res)
 }
 
+pub(crate) fn handle_test_runnables_in_db(
```

Comment:
I believe this function is exactly what we do not want to support? Judging from https://github.com/rust-lang/rust-analyzer/issues/3601#issuecomment-945778285 and https://github.com/rust-lang/rust-analyzer/issues/3601#issuecomment-960509767 though I am also not too knowledgeable about this issue.

Reply:
Agree, I am actually kind of confused here.
At least, we need to know the entry for each target(.i.e, the binary files, the integration files, the library root...). This needs the knowledge of the build tool(.e.g, cargo)
BSP seems to be a good way to support this feature(to know the structure of the project, not to know the tests of the project, BSP seems to support it https://build-server-protocol.github.io/docs/specification#test-request, but it's kind of weird for me). But I am not sure.

Anyway, I will try to add another API as https://github.com/rust-lang/rust-analyzer/issues/3601#issuecomment-945778285 said firstly.

Reply:
Auh, I would say this is kind of blocked by BSP?

If we finally have BSP, would we export the APIs directly? And if so, could I implment two of them firstly in this PR?
I think these two would be enough: `workspace/buildTargets` and `buildTarget/sources`.

And in the future, if we finally have BSP, we could redirect the request to the real BSP server rather than the hack  one.

Reply:
Friendly ping @matklad to get any opinion...

Reply:
Yup, Veykril is correct here: we do not want “get all the tests everywhere” functionality, as it is fundamentally at odds with lazy analysis.

We should export two API:

* get all runnables in a file (already exists as runnables)
* get available test targets (as in “Cargo targets”), this one is to be implemented.

I don’t think this needs to be blocked on BSP: BSP solves a particular problem of Java/Scala ecosystem that there’s no dominant build tool. For rust, exposing Cargo metadata as is should be fine. I don’t think we need any support for rust-project.json here. That one deliberately lacks any info about how to actually run stuff, and, eg, integrating runnables with buck should be on bucks’s extension.

Reply:
After three years since #3601 was opened, I think O(n) is better than O(never).

Reply:
I am refactoring the code slowly to achieve the goal, I guess 70% is finished.

It's a good signal for me that the members could hold on their opinion, which will lead to a well coded program! And a good program will attract more people to contribute.

So, please wait for some other days, after all, it has been 3 years :P

Reply:
> After three years since https://github.com/rust-lang/rust-analyzer/issues/3601 was opened, I think O(n) is better than O(never).

@coder0xff I think a lot of features can be argued for under that framing. That being said, one of the most attractive points of rust-analyzer is how _fast_ it is. An `O(n)` solution fundamentally compromises one of rust-analyzer's fundamental design tenets, and given that there are approaches that do not require compromising said tenets (and, as a consequence, speed), I'd argue that no solution is better than the wrong solution.

> I am refactoring the code slowly to achieve the goal, I guess 70% is finished.

@ShuiRuTian What goal are you referring to here? Is it the lazy approach, or...?

Reply:
Oh, sorry, somehow I miss this message....

Yes, the refactoring will be it more easier to change. Although to be totally laziness should be the last step of the refactoring, I am fixing bugs and trying to provide some better experience.